### PR TITLE
Fix contained associations not having types mapped.

### PR DESCRIPTION
**Pull request by markstory**
When contain() is called, use the normalized information to set type mappings. This forces associations to be normalized earlier, but allows the ORM to be more consistent and not require developers to set type information.

Refs #6975

### CODE REVIEW

**Comment by lorenzo** [comment]
Out of curiosity, could you benchmark the impact of this change when binding a query and calling contain() multiple times?

I feel like calling addDefaultTypes so many times (twice per association) could slow down things a bit.

**Comment by markstory** [comment]
> I feel like calling addDefaultTypes so many times (twice per association) could slow down things a bit.

It will only be called as many times as you call contain. I'll run the numbers though. Does something like timing a query with 5 contained associations seem reasonable?

**Comment by lorenzo** [comment]
No, it is done twice. Once for the main query (your change) and once for the surrogate query when calling attachTo.

The testing strategy seems reasonable to me. Just make sure you call contain multiple types instead of a single array.

**Comment by markstory** [comment]
So the performance delta is significant. For the following shell task

```php
        $bookmarks = TableRegistry::get('Bookmarks');
        $start = microtime(true);
        foreach (range(0, 10000) as $i) {
            $q = $bookmarks->find()
                ->contain('Tags')
                ->contain('Users');
        }
        $end = microtime(true);
        $this->out('Buliding 10000 queries took ' . ($end - $start) . 'seconds');

```

The timing goes from ~0.35 seconds, to 1.08 seconds. I had the metadata cache enabled, and xdebug disabled. I'll try only adding type data for 'new' associations, and see if that has less of an impact.

**Comment by markstory** [comment]
@lorenzo Updated so the test code only takes ~0.65 seconds. I've included a test ensuring nested associations work as well.

**Comment by lorenzo** [comment]
Nice job!

File: src/ORM/Query.php [hunk 1]
```diff
@@ -288,22 +287,51 @@ public function eagerLoader(EagerLoader $instance = null)
...
+    {
+        $typeMap = $this->typeMap();
+        foreach ($associations as $name => $nested) {
+            $association = $table->association($name);
+            if (!$association) {
+                continue;
+            }
+            $target = $association->target();
+            $primary = (array)$target->primaryKey();
+            if ($typeMap->type($target->aliasField($primary[0])) === null) {
```
Comment (ADmad) [inline src/ORM/Query.php]: This line generates a "undefined index" error if table doesn't have a primary key, e.g. for a join table. Everything works fine when using a regular belongsToMany between 2 models but if you set an association with join table's model you have the error.

Comment (lorenzo) [inline src/ORM/Query.php]: There are actually many things that will fail when not having a primary key in the join table... For example the Marshallese will not be able to work without it. Have you been using a table like that for a long time now?

Comment (ADmad) [inline src/ORM/Query.php]: Just started using this table. This is an old 2.x project for which I am making new frontend in 3.x. There are other join tables without primary key but until now haven't needed to setup an association with join table's model.

I also realize now that for whatever models with belongsToMany associations I have done writes to have had join tables with extra fields and primary key field, which is why I have never had a problem with marshaling either.

Comment (lorenzo) [inline src/ORM/Query.php]: Ok, do you want to get a fix done for this? In the meantime I suggest adding the composite primary key to the table

Comment (ADmad) [inline src/ORM/Query.php]: Yeah will open a PR for this. Already added composite primary key to the table before commenting here :)

**Pull request merged**